### PR TITLE
WIP :Screencast

### PR DESCRIPTION
**Pull request by Decodetalkers**
Emm. try to make one

### CODE REVIEW

File: src/screencast.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    #[dbus_interface(property)]
+    fn available_cursor_modes(&self) -> u32 {
+        (CursorMode::Hidden | CursorMode::Embedded).bits()
+    }
+
+    #[dbus_interface(property)]
+    fn available_source_types(&self) -> u32 {
+        BitFlags::from_flag(SourceType::Monitor).bits()
```
Comment (Decodetalkers) [inline src/screencast.rs:75]: I want to support region cast later.., so  virtual maybe later? but what about window?

File: src/screencast.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    ) -> zbus::fdo::Result<(u32, StartReturnValue)> {
+        let locked_sessions = SESSIONS.lock().await;
+        let Some(index) = locked_sessions.iter().position(|this_session| this_session.handle_path == session_handle.clone().into()) else {
+            tracing::warn!("No session is created or it is removed");
+            return Ok((2, StartReturnValue::default()));
+        };
+        let current_session = locked_sessions[index].clone();
+        drop(locked_sessions);
+        let show_cursor = current_session.cursor_mode.show_cursor();
+        println!("{:?}", current_session);
```
Comment (Decodetalkers) [inline src/screencast.rs:142]: here will leave a stream, with its fd id..from the doc and read xdg-desktop-portal-kde, and the part of wlr

File: src/screencast.rs (hunk 1 above)
Comment (Shinyzenith) [reply src/screencast.rs:75]: Window cast will need top-level data from compositor. We should look into making our own Wayland protocol to expose these details in that case.

File: src/screencast.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+#[zvariant(signature = "dict")]
+pub struct SelectSourcesOptions {
+    /// A string that will be used as the last element of the handle.
+    /// What types of content to record.
+    pub types: Option<BitFlags<SourceType>>,
+    /// Whether to allow selecting multiple sources.
+    pub multiple: Option<bool>,
+    /// Determines how the cursor will be drawn in the screen cast stream.
+    pub cursor_mode: Option<CursorMode>,
+    pub restore_token: Option<String>,
```
Comment (Decodetalkers) [inline src/screencast.rs:32]: the key is a little different from xdg-desktop-portal document, but the code is from ashpd.. so maybe they are wrong? but it is optional, and we do not need it now, so it is safe

File: src/screencast.rs (hunk 2 above)
Comment (Decodetalkers) [reply src/screencast.rs:142]: I want there to be a dialog, but this maybe some discussion. what to show

**Comment by Decodetalkers** [comment]
so, for the next development, I will merge it now

**Pull request merged**